docs/designs/allreduce_based_horovod.md

```diff
@@ -0,0 +1,117 @@
+# Design for Elastic AllReduce-based Training Support Based on Horovod
+
+This document describes the design for supporting AllReduce-based distributed
+training based on Horovod in ElasticDL.
+
+## Motivation
+
+We have developed elastic AllReduce based on FTlib in ElasticDL.
+From the [benchmark report](../benchmark/ftlib_benchmark.md), we can
+find that the performance of FTlib for ResNet50 is worse than Horovod.
+What's more, it is not stable to build consensus by gossip protocol in FTlib
+like the [issue](https://github.com/sql-machine-learning/elasticdl/issues/2192#issuecomment-664096185).
+
+FTlib uses Gloo to implement elastic AllReduce because the worker process
+can catch the exception from Gloo and not exit if the AllReduce operator
+fails. Horovod also can use Gloo as backend. What's more, there are many
+optimizations like [Tensor Fusion](https://horovod.readthedocs.io/en/latest/tensor-fusion_include.html)
+in Horovod to improve performance. So, the performance of Horovod for ResNet50 is
+better than FTlib because ResNet50 has may small tensors.
+
+Horovod provides Python APIs for Gloo and we can use those APIs in ElasticDL
+to implement elastic AllReduce.
+
+## ElasticDL Re-initialize Horovod When the Number of Workers Changes
+
+When using Horovod with Gloo backend, we need to create a `RendezvousServer`
+and put worker hosts into the KVStore implemented by HTTP for Gloo.
+In ElasticDL, there is a master to manage workers. The master can get
+all worker hosts and create a `RendezvousServer`. After the master launches
+workers, it can set worker hosts into KVStore of Horovod for Gloo.
+
+```python
+import horovod
+from horovod.run.http.http_server import  RendezvousServer
+from horovod.runner.common.util.hosts import get_host_assignments
+
+hosts = get_worker_hosts()
+
+host_alloc_plan = get_host_assignments(hosts, num_proc)
+global_rendezv_port = rendezvous.start()
+
+# Set hosts into KVStore for Gloo
+rendezvous.init(host_alloc_plan)
+```
+
+Then, the worker can call `hvd.init` to initialize the Gloo context for
+AllReduce.
+
+When the master finds the number of workers changes, it can re-create a new
+`RendezvousServer` and notify workers to re-initialize Horovod.
+In the Kubernetes cluster, the number of workers may change for the
+following reasons:
+
+1. Some workers fail because of preemption.
+1. The master re-launches new workers.
```

Comment:
A worker pod status becomes running. 
(master re-launches new workers  != worker pod is up and running)